### PR TITLE
Support request redirects in Page Graph.

### PR DESCRIPTION
**Pull request by goodov**
#### Tracking resources redirects

The graph now accurately depicts request redirects. When a redirect is
identified, resource nodes can now establish connections to other resource nodes
using the `request redirect` edge.

Resolves https://github.com/brave/brave-browser/issues/35867

https://github.com/brave/brave-core/assets/5928869/728bf5b0-f087-4fb1-a456-881bde915bcb

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by goodov** [comment]
this should be merged after https://github.com/brave/brave-core/pull/22335

File: third_party/blink/renderer/core/brave_page_graph/requests/request_tracker.h [hunk 1]
```diff
@@ -81,6 +87,8 @@ class RequestTracker {
   scoped_refptr<const TrackedRequestRecord> ReturnTrackingRecord(
       const InspectorId request_id);
 
+  PageGraphContext* const page_graph_context_ = nullptr;
```
[the remark this answers is not in the document]
Comment (goodov) [reply third_party/blink/renderer/core/brave_page_graph/requests/request_tracker.h:90]: this is renderer code. `raw_ptr` is not required in renderer code (at least for now).

File: third_party/blink/renderer/core/brave_page_graph/requests/tracked_request.h [hunk 2]
```diff
@@ -51,7 +57,12 @@ class TrackedRequest {
  protected:
   void FinishResponseBodyHash();
 
-  enum class RequestStatus : uint8_t { kError = 0, kSuccess, kUnknown };
+  enum class RequestStatus {
+    kError,
+    kSuccess,
+  };
+
+  PageGraphContext* const page_graph_context_ = nullptr;
```
[the remark this answers is not in the document]
Comment (goodov) [reply third_party/blink/renderer/core/brave_page_graph/requests/tracked_request.h:65]: this is renderer code. `raw_ptr` is not required in renderer code (at least for now).

**Comment by pes10k** [comment]
This looks fantastic, thank you @goodov !

File: third_party/blink/renderer/core/brave_page_graph/requests/request_tracker.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (fmarier) [reply third_party/blink/renderer/core/brave_page_graph/requests/request_tracker.h:90]: https://github.com/brave/security-action/issues/548 for excluding `third_party/blink/renderer/*`

**Comment by pes10k** [comment]
@goodov @fmarier just a ping on this, since it looks like the sec review is complete.  Is this good to go?

**Comment by fmarier** [comment]
Yes, the security alert was a false positive. This is ready for a code review from the appropriate codeowners.

File: third_party/blink/renderer/core/brave_page_graph/requests/tracked_request.cc [hunk 3]
```diff
@@ -16,25 +21,30 @@ namespace brave_page_graph {
...
-      resource_type_(resource_type),
-      resource_(resource) {
-  CHECK(requester != nullptr);
-  CHECK(resource != nullptr);
+    : page_graph_context_(page_graph_context),
+      request_id_(request_id),
+      resource_type_(resource_type) {
+  DCHECK(page_graph_context_);
+  DCHECK(requester);
+  DCHECK(resource);
```
Comment (iefremov) [inline third_party/blink/renderer/core/brave_page_graph/requests/tracked_request.cc:34]: why CHECK is now DCHECK

**Review by iefremov — APPROVED** [review]

File: third_party/blink/renderer/core/brave_page_graph/requests/tracked_request.cc (hunk 3 above)
Comment (goodov) [reply third_party/blink/renderer/core/brave_page_graph/requests/tracked_request.cc:34]: these are mostly sanity checks, not runtime checks.

**Pull request merged**